File: swf-parser.ts/src/lib/parsers/shapes.ts
import {Sint32, Uint16, UintSize} from "semantic-types";
import {
  CapStyle, FillStyle, fillStyles, FillStyleType, Fixed8P8, Glyph, Gradient, JoinStyleType, LineStyle, Matrix, Shape,
  ShapeRecord,
  shapeRecords, ShapeRecordType, SRgb8, StraightSRgba8, Vector2D,
} from "swf-tree";
import {BitStream, ByteStream} from "../stream";
import {parseMatrix, parseSRgb8, parseStraightSRgba8} from "./basic-data-types";
import {parseGradient} from "./gradient";

export enum ShapeVersion {
  Shape1,
  Shape2,
  Shape3,
}

export function parseGlyph(byteStream: ByteStream): Glyph {
  const bitStream: BitStream = byteStream.asBitStream();
  const result: Glyph = parseGlyphBits(bitStream);
  bitStream.align();
  return result;
}

export function parseGlyphBits(bitStream: BitStream): Glyph {
  const fillBits: UintSize = bitStream.readUint32Bits(4);
  const lineBits: UintSize = bitStream.readUint32Bits(4);
  const records: ShapeRecord[] = parseShapeRecordStringBits(bitStream, fillBits, lineBits, ShapeVersion.Shape1);
  return {records};
}

export function parseShape(byteStream: ByteStream, version: ShapeVersion): Shape {
  const bitStream: BitStream = byteStream.asBitStream();
  const result: Shape = parseShapeBits(bitStream, version);
  bitStream.align();
  return result;
}

export function parseShapeBits(bitStream: BitStream, version: ShapeVersion): Shape {
  const styles: ShapeStyles = parseShapeStylesBits(bitStream, version);
  const records: ShapeRecord[] = parseShapeRecordStringBits(
    bitStream,
    styles.fillBits,
    styles.lineBits,
    version,
  );
  return {
    fillStyles: styles.fill,
    lineStyles: styles.line,
    records,
  };
}

export interface ShapeStyles {
  fill: FillStyle[];
  line: LineStyle[];
  fillBits: UintSize;
  lineBits: UintSize;
}

export function parseShapeStylesBits(bitStream: BitStream, version: ShapeVersion): ShapeStyles {
  const byteStream: ByteStream = bitStream.asByteStream();
  const fill: FillStyle[] = parseFillStyleList(byteStream, version);
  const line: LineStyle[] = parseLineStyleList(byteStream, version);
  bitStream = byteStream.asBitStream();
  const fillBits: UintSize = bitStream.readUint32Bits(4);
  const lineBits: UintSize = bitStream.readUint32Bits(4);
  return {fill, line, fillBits, lineBits};
}

export function parseShapeRecordStringBits(
  bitStream: BitStream,
  fillBits: UintSize,
  lineBits: UintSize,
  version: ShapeVersion,
): ShapeRecord[] {
  const result: ShapeRecord[] = [];

  while (true) {
    const bytePos: number = bitStream.bytePos;
    const bitPos: number = bitStream.bitPos;
    const head: number = bitStream.readUint16Bits(6);
    if (head === 0) {
      break;
    } else {
      bitStream.bytePos = bytePos;
      bitStream.bitPos = bitPos;
    }

    const isEdge: boolean = bitStream.readBoolBits();
    if (isEdge) {
      const isStraightEdge: boolean = bitStream.readBoolBits();
      if (isStraightEdge) {
        result.push(parseStraightEdgeBits(bitStream));
      } else {
        result.push(parseCurvedEdgeBits(bitStream));
      }
    } else {
      let styles: shapeRecords.StyleChange;
      [styles, [fillBits, lineBits]] = parseStyleChangeBits(bitStream, fillBits, lineBits, version);
      result.push(styles);
    }
  }

  return result;
}

export function parseCurvedEdgeBits(bitStream: BitStream): shapeRecords.CurvedEdge {
  const nBits: UintSize = bitStream.readUint16Bits(4) + 2;
  const controlX: Sint32 = bitStream.readSint32Bits(nBits);
  const controlY: Sint32 = bitStream.readSint32Bits(nBits);
  const deltaX: Sint32 = bitStream.readSint32Bits(nBits);
  const deltaY: Sint32 = bitStream.readSint32Bits(nBits);
  return {
    type: ShapeRecordType.CurvedEdge,
    controlDelta: {x: controlX, y: controlY},
    endDelta: {x: deltaX, y: deltaY},
  };
}

export function parseStraightEdgeBits(bitStream: BitStream): shapeRecords.StraightEdge {
  const nBits: UintSize = bitStream.readUint16Bits(4) + 2;
  const isDiagonal: boolean = bitStream.readBoolBits();
  const isVertical: boolean = !isDiagonal && bitStream.readBoolBits();
  const deltaX: Sint32 = isDiagonal || !isVertical ? bitStream.readSint32Bits(nBits) : 0;
  const deltaY: Sint32 = isDiagonal || isVertical ? bitStream.readSint32Bits(nBits) : 0;
  return {
    type: ShapeRecordType.StraightEdge,
    endDelta: {x: deltaX, y: deltaY},
  };
}

export function parseStyleChangeBits(
  bitStream: BitStream,
  fillBits: UintSize,
  lineBits: UintSize,
  version: ShapeVersion,
): [shapeRecords.StyleChange, [UintSize, UintSize]] {
  const hasNewStyles: boolean = bitStream.readBoolBits();
  const changeLineStyle: boolean = bitStream.readBoolBits();
  const changeRightFill: boolean = bitStream.readBoolBits();
  const changeLeftFill: boolean = bitStream.readBoolBits();
  const hasMoveTo: boolean = bitStream.readBoolBits();

  let moveTo: Vector2D | undefined = undefined;
  if (hasMoveTo) {
    const nBits: UintSize = bitStream.readUint16Bits(5);
    const x: Sint32 = bitStream.readSint32Bits(nBits);
    const y: Sint32 = bitStream.readSint32Bits(nBits);
    moveTo = {x, y};
  }
  const leftFill: UintSize | undefined = changeLeftFill ? bitStream.readUint16Bits(fillBits) : undefined;
  const rightFill: UintSize | undefined = changeRightFill ? bitStream.readUint16Bits(fillBits) : undefined;
  const lineStyle: UintSize | undefined = changeLineStyle ? bitStream.readUint16Bits(lineBits) : undefined;

  let fillStyles: FillStyle[] | undefined = undefined;
  let lineStyles: LineStyle[] | undefined = undefined;
  if (hasNewStyles) {
    const styles: ShapeStyles = parseShapeStylesBits(bitStream, version);
    fillStyles = styles.fill;
    lineStyles = styles.line;
    fillBits = styles.fillBits;
    lineBits = styles.lineBits;
  }

  const styleChangeRecord: shapeRecords.StyleChange = {
    type: ShapeRecordType.StyleChange,
    moveTo,
    leftFill,
    rightFill,
    lineStyle,
    fillStyles,
    lineStyles,
  };

  return [styleChangeRecord, [fillBits, lineBits]];
}

/**
 * Parse a fill style list length or line style list length.
 *
 * @param byteStream Stream to use to parse this list length. Will mutate its state.
 * @param allowExtended Allow extended size (`> 255`). Here are the recommended values:
 *                      - `true` for `DefineShape2`, `DefineShape3`
 *                      - `false` for `DefineShape`
 * @returns List length
 */
export function parseListLength(byteStream: ByteStream, allowExtended: boolean): UintSize {
  const len: UintSize = byteStream.readUint8();
  if (len === 0xff && allowExtended) {
    return byteStream.readUint16LE();
  } else {
    return len;
  }
}

export function parseFillStyle(byteStream: ByteStream, withAlpha: boolean): FillStyle {
  switch (byteStream.readUint8()) {
    case 0x00:
      return parseSolidFill(byteStream, withAlpha);
    case 0x10:
      return parseLinearGradientFill(byteStream, withAlpha);
    case 0x12:
      return parseRadialGradientFill(byteStream, withAlpha);
    case 0x13:
      return parseFocalGradientFill(byteStream, withAlpha);
    case 0x40:
      return parseBitmapFill(byteStream, true, true);
    case 0x41:
      return parseBitmapFill(byteStream, false, true);
    case 0x42:
      return parseBitmapFill(byteStream, true, false);
    case 0x43:
      return parseBitmapFill(byteStream, false, false);
    default:
      throw new Error("Unexpected fill style");
  }
}

export function parseBitmapFill(byteStream: ByteStream, repeating: boolean, smoothed: boolean): fillStyles.Bitmap {
  const bitmapId: Uint16 = byteStream.readUint16LE();
  const matrix: Matrix = parseMatrix(byteStream);
  return {
    type: FillStyleType.Bitmap,
    bitmapId,
    matrix,
    repeating,
    smoothed,
  };
}

export function parseFocalGradientFill(byteStream: ByteStream, withAlpha: boolean): fillStyles.FocalGradient {
  const matrix: Matrix = parseMatrix(byteStream);
  const gradient: Gradient = parseGradient(byteStream, withAlpha);
  const focalPoint: Fixed8P8 = byteStream.readFixed8P8LE();
  return {
    type: FillStyleType.FocalGradient,
    matrix,
    gradient,
    focalPoint,
  };
}

export function parseLinearGradientFill(byteStream: ByteStream, withAlpha: boolean): fillStyles.LinearGradient {
  const matrix: Matrix = parseMatrix(byteStream);
  const gradient: Gradient = parseGradient(byteStream, withAlpha);
  const focalPoint: Fixed8P8 = byteStream.readFixed8P8LE();
  return {
    type: FillStyleType.LinearGradient,
    matrix,
    gradient,
  };
}

export function parseRadialGradientFill(byteStream: ByteStream, withAlpha: boolean): fillStyles.RadialGradient {
  const matrix: Matrix = parseMatrix(byteStream);
  const gradient: Gradient = parseGradient(byteStream, withAlpha);
  const focalPoint: Fixed8P8 = byteStream.readFixed8P8LE();
  return {
    type: FillStyleType.RadialGradient,
    matrix,
    gradient,
  };
}

export function parseSolidFill(byteStream: ByteStream, withAlpha: boolean): fillStyles.Solid {
  let color: StraightSRgba8;
  if (withAlpha) {
    color = parseStraightSRgba8(byteStream);
  } else {
    color = {...parseSRgb8(byteStream), a: 255};
  }
  return {
    type: FillStyleType.Solid,
    color,
  };
}

export function parseFillStyleList(
  byteStream: ByteStream,
  version: ShapeVersion,
): FillStyle[] {
  const result: FillStyle[] = [];
  const len: UintSize = parseListLength(byteStream, version !== ShapeVersion.Shape1);
  for (let i: UintSize = 0; i < len; i++) {
    result.push(parseFillStyle(byteStream, version === ShapeVersion.Shape3));
  }
  return result;
}

export function parseLineStyle(byteStream: ByteStream): LineStyle {
  const width: Uint16 = byteStream.readUint16LE();
  const color: SRgb8 = parseSRgb8(byteStream);
  return {
    width,
    startCap: CapStyle.Round,
    endCap: CapStyle.Round,
    join: {type: JoinStyleType.Round},
    noHScale: false,
    noVScale: false,
    noClose: false,
    pixelHinting: false,
    fill: {
      type: FillStyleType.Solid,
      color: {...color, a: 255},
    },
  };
}

export function parseLineStyleList(
  byteStream: ByteStream,
  version: ShapeVersion,
): LineStyle[] {
  const result: LineStyle[] = [];
  const len: UintSize = parseListLength(byteStream, version !== ShapeVersion.Shape1);
  for (let i: UintSize = 0; i < len; i++) {
    result.push(parseLineStyle(byteStream));
  }
  return result;
}

File: swf-parser.rs/src/parsers/shapes.rs
use swf_tree as ast;
use nom::{IResult, Needed};
use nom::{le_u8 as parse_u8, le_u16 as parse_le_u16};
use parsers::basic_data_types::{
  parse_bool_bits,
  parse_i32_bits,
  parse_u16_bits,
  parse_u32_bits,
  parse_le_fixed8_p8,
  parse_matrix,
  parse_s_rgb8,
  parse_straight_s_rgba8
};
use parsers::gradient::parse_gradient;

#[derive(PartialEq, Eq, Clone, Copy)]
pub enum ShapeVersion {
  Shape1,
  Shape2,
  Shape3,
  //  Shape4,
}

pub fn parse_glyph(input: &[u8]) -> IResult<&[u8], ast::Glyph> {
  bits!(input, parse_glyph_bits)
}

pub fn parse_glyph_bits(input: (&[u8], usize)) -> IResult<(&[u8], usize), ast::Glyph> {
  do_parse!(
    input,
    fill_bits: map!(apply!(parse_u32_bits, 4), |x| x as usize) >>
    line_bits: map!(apply!(parse_u32_bits, 4), |x| x as usize) >>
    // TODO: Check which shape version to use
    records: apply!(parse_shape_record_string_bits, fill_bits, line_bits, ShapeVersion::Shape1) >>
    (ast::Glyph {
      records: records,
    })
  )
}

pub fn parse_shape(input: &[u8], version: ShapeVersion) -> IResult<&[u8], ast::Shape> {
  bits!(input, apply!(parse_shape_bits, version))
}

pub fn parse_shape_bits(input: (&[u8], usize), version: ShapeVersion) -> IResult<(&[u8], usize), ast::Shape> {
  do_parse!(
    input,
    styles: apply!(parse_shape_styles_bits, version) >>
    records: apply!(parse_shape_record_string_bits, styles.fill_bits, styles.line_bits, version) >>
    (ast::Shape {
      fill_styles: styles.fill,
      line_styles: styles.line,
      records: records,
    })
  )
}

pub struct ShapeStyles {
  pub fill: Vec<ast::FillStyle>,
  pub line: Vec<ast::LineStyle>,
  pub fill_bits: usize,
  pub line_bits: usize,
}

pub fn parse_shape_styles_bits(input: (&[u8], usize), version: ShapeVersion) -> IResult<(&[u8], usize), ShapeStyles> {
  do_parse!(
    input,
    fill: bytes!(apply!(parse_fill_style_list, version)) >>
    line: bytes!(apply!(parse_line_style_list, version)) >>
    fill_bits: map!(apply!(parse_u32_bits, 4), |x| x as usize) >>
    line_bits: map!(apply!(parse_u32_bits, 4), |x| x as usize) >>
    (ShapeStyles {
      fill: fill,
      line: line,
      fill_bits: fill_bits,
      line_bits: line_bits,
    })
  )
}

pub fn parse_shape_record_string_bits(input: (&[u8], usize), mut fill_bits: usize, mut line_bits: usize, version: ShapeVersion) -> IResult<(&[u8], usize), Vec<ast::ShapeRecord>> {
  let mut result: Vec<ast::ShapeRecord> = Vec::new();
  let mut current_input = input;

  loop {
    match parse_u16_bits(current_input, 6) {
      IResult::Done(next_input, record_head) => if record_head == 0 {
        current_input = next_input;
        break
      },
      IResult::Error(e) => return IResult::Error(e),
      IResult::Incomplete(_) => return IResult::Incomplete(Needed::Unknown),
    };

    let is_edge = match parse_bool_bits(current_input) {
      IResult::Done(next_input, is_edge) => {
        current_input = next_input;
        is_edge
      }
      IResult::Error(e) => return IResult::Error(e),
      IResult::Incomplete(_) => return IResult::Incomplete(Needed::Unknown),
    };

    if is_edge {
      let is_straight_edge = match parse_bool_bits(current_input) {
        IResult::Done(next_input, is_straight_edge) => {
          current_input = next_input;
          is_straight_edge
        }
        IResult::Error(e) => return IResult::Error(e),
        IResult::Incomplete(_) => return IResult::Incomplete(Needed::Unknown),
      };
      if is_straight_edge {
        match parse_straight_edge_bits(current_input) {
          IResult::Done(next_input, straight_edge) => {
            let record = ast::ShapeRecord::StraightEdge(straight_edge);
            result.push(record);
            current_input = next_input;
          }
          IResult::Error(e) => return IResult::Error(e),
          IResult::Incomplete(n) => return IResult::Incomplete(n),
        };
      } else {
        match parse_curved_edge_bits(current_input) {
          IResult::Done(next_input, curved_edge) => {
            let record = ast::ShapeRecord::CurvedEdge(curved_edge);
            result.push(record);
            current_input = next_input;
          }
          IResult::Error(e) => return IResult::Error(e),
          IResult::Incomplete(n) => return IResult::Incomplete(n),
        };
      }
    } else {
      match parse_style_change_bits(current_input, fill_bits, line_bits, version) {
        IResult::Done(next_input, record_and_bits) => {
          let (style_change, style_bits) = record_and_bits;
          fill_bits = style_bits.0;
          line_bits = style_bits.1;
          let record = ast::ShapeRecord::StyleChange(style_change);
          result.push(record);
          current_input = next_input;
        }
        IResult::Error(e) => return IResult::Error(e),
        IResult::Incomplete(n) => return IResult::Incomplete(n),
      };
    }
  }

  IResult::Done(current_input, result)
}

pub fn parse_curved_edge_bits(input: (&[u8], usize)) -> IResult<(&[u8], usize), ast::shape_records::CurvedEdge> {
  do_parse!(
    input,
    n_bits: map!(apply!(parse_u16_bits, 4), |x| x as usize) >>
    control_x: apply!(parse_i32_bits, n_bits + 2) >>
    control_y: apply!(parse_i32_bits, n_bits + 2) >>
    delta_x: apply!(parse_i32_bits, n_bits + 2) >>
    delta_y: apply!(parse_i32_bits, n_bits + 2) >>
    (ast::shape_records::CurvedEdge {
      control_delta: ast::Vector2D {x: control_x, y: control_y},
      anchor_delta: ast::Vector2D {x: delta_x, y: delta_y},
    })
  )
}

pub fn parse_straight_edge_bits(input: (&[u8], usize)) -> IResult<(&[u8], usize), ast::shape_records::StraightEdge> {
  do_parse!(
    input,
    n_bits: map!(apply!(parse_u16_bits, 4), |x| x as usize) >>
    is_diagonal: call!(parse_bool_bits) >>
    is_vertical: map!(cond!(!is_diagonal, call!(parse_bool_bits)), |opt: Option<bool>| opt.unwrap_or_default()) >>
    delta_x: cond!(is_diagonal || !is_vertical, apply!(parse_i32_bits, n_bits + 2)) >>
    delta_y: cond!(is_diagonal || is_vertical, apply!(parse_i32_bits, n_bits + 2)) >>
    (ast::shape_records::StraightEdge {
      delta: ast::Vector2D {x: delta_x.unwrap_or_default(), y: delta_y.unwrap_or_default()},
    })
  )
}

pub fn parse_style_change_bits(input: (&[u8], usize), fill_bits: usize, line_bits: usize, version: ShapeVersion) -> IResult<(&[u8], usize), (ast::shape_records::StyleChange, (usize, usize))> {
  do_parse!(
    input,
    has_new_styles: parse_bool_bits >>
    change_line_style: call!(parse_bool_bits) >>
    change_right_fill: call!(parse_bool_bits) >>
    change_left_fill: call!(parse_bool_bits) >>
    has_move_to: call!(parse_bool_bits) >>
    move_to: cond!(has_move_to,
      do_parse!(
        move_to_bits: apply!(parse_u16_bits, 5) >>
        x: apply!(parse_i32_bits, move_to_bits as usize) >>
        y: apply!(parse_i32_bits, move_to_bits as usize) >>
        (ast::Vector2D {x: x, y: y})
      )
    ) >>
    left_fill: cond!(change_left_fill, apply!(parse_u16_bits, fill_bits)) >>
    right_fill: cond!(change_right_fill, apply!(parse_u16_bits, fill_bits)) >>
    line_style: cond!(change_line_style, apply!(parse_u16_bits, line_bits)) >>
    styles: map!(
      cond!(has_new_styles, apply!(parse_shape_styles_bits, version)),
      |styles| match styles {
        Option::Some(styles) => (Option::Some(styles.fill), Option::Some(styles.line), styles.fill_bits, styles.line_bits),
        Option::None => (Option::None, Option::None, fill_bits, line_bits),
      }
    ) >>
    ((
      ast::shape_records::StyleChange {
          move_to: move_to,
          left_fill: left_fill.map(|x| x as usize),
          right_fill: right_fill.map(|x| x as usize),
          line_style: line_style.map(|x| x as usize),
          fill_styles: styles.0,
          line_styles: styles.1,
      },
      (styles.2, styles.3),
    ))
  )
}

pub fn parse_list_length(input: &[u8], support_extended: bool) -> IResult<&[u8], usize> {
  match parse_u8(input) {
    IResult::Done(remaining_input, u8_len) => {
      if u8_len == 0xff && support_extended {
        parse_le_u16(remaining_input).map(|x| x as usize)
      } else {
        IResult::Done(remaining_input, u8_len as usize)
      }
    }
    IResult::Error(e) => IResult::Error(e),
    IResult::Incomplete(n) => IResult::Incomplete(n),
  }
}

pub fn parse_fill_style_list(input: &[u8], version: ShapeVersion) -> IResult<&[u8], Vec<ast::FillStyle>> {
  length_count!(input, apply!(parse_list_length, version != ShapeVersion::Shape1), apply!(parse_fill_style, version == ShapeVersion::Shape3))
}

#[allow(unused_variables)]
pub fn parse_fill_style(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::FillStyle> {
  switch!(input, parse_u8,
    0x00 => map!(apply!(parse_solid_fill, with_alpha), |fill| ast::FillStyle::Solid(fill)) |
    0x10 => map!(apply!(parse_linear_gradient_fill, with_alpha), |fill| ast::FillStyle::LinearGradient(fill)) |
    0x12 => map!(apply!(parse_radial_gradient_fill, with_alpha), |fill| ast::FillStyle::RadialGradient(fill)) |
    0x13 => map!(apply!(parse_focal_gradient_fill, with_alpha), |fill| ast::FillStyle::FocalGradient(fill)) |
    0x40 => map!(apply!(parse_bitmap_fill, true, true), |fill| ast::FillStyle::Bitmap(fill)) |
    0x41 => map!(apply!(parse_bitmap_fill, false, true), |fill| ast::FillStyle::Bitmap(fill)) |
    0x42 => map!(apply!(parse_bitmap_fill, true, false), |fill| ast::FillStyle::Bitmap(fill)) |
    0x43 => map!(apply!(parse_bitmap_fill, false, false), |fill| ast::FillStyle::Bitmap(fill))
    // TODO: Error
  )
}

pub fn parse_bitmap_fill(input: &[u8], repeating: bool, smoothed: bool) -> IResult<&[u8], ast::fill_styles::Bitmap> {
  do_parse!(
    input,
    bitmap_id: parse_le_u16 >>
    matrix: parse_matrix >>
    (ast::fill_styles::Bitmap {
      bitmap_id: bitmap_id,
      matrix: matrix,
      repeating: repeating,
      smoothed: smoothed
    })
  )
}

pub fn parse_focal_gradient_fill(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::fill_styles::FocalGradient> {
  do_parse!(
    input,
    matrix: parse_matrix >>
    gradient: apply!(parse_gradient, with_alpha) >>
    focal_point: parse_le_fixed8_p8 >>
    (ast::fill_styles::FocalGradient {
      matrix: matrix,
      gradient: gradient,
      focal_point: focal_point
    })
  )
}

pub fn parse_linear_gradient_fill(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::fill_styles::LinearGradient> {
  do_parse!(
    input,
    matrix: parse_matrix >>
    gradient: apply!(parse_gradient, with_alpha) >>
    (ast::fill_styles::LinearGradient {
      matrix: matrix,
      gradient: gradient
    })
  )
}

pub fn parse_radial_gradient_fill(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::fill_styles::RadialGradient> {
  do_parse!(
    input,
    matrix: parse_matrix >>
    gradient: apply!(parse_gradient, with_alpha) >>
    (ast::fill_styles::RadialGradient {
      matrix: matrix,
      gradient: gradient
    })
  )
}

#[allow(unused_variables)]
pub fn parse_solid_fill(input: &[u8], with_alpha: bool) -> IResult<&[u8], ast::fill_styles::Solid> {
  do_parse!(
    input,
    color: switch!(value!(with_alpha),
      true => call!(parse_straight_s_rgba8) |
      false => map!(parse_s_rgb8, |c| ast::StraightSRgba8 {r: c.r, g: c.g, b: c.b, a: 255})
    ) >>
    (ast::fill_styles::Solid {color: color})
  )
}

pub fn parse_line_style_list(input: &[u8], version: ShapeVersion) -> IResult<&[u8], Vec<ast::LineStyle>> {
  length_count!(input, apply!(parse_list_length, version != ShapeVersion::Shape1), parse_line_style)
}

pub fn parse_line_style(input: &[u8]) -> IResult<&[u8], ast::LineStyle> {
  do_parse!(
    input,
    width: parse_le_u16 >>
    color: parse_s_rgb8 >>
    (
      ast::LineStyle {
      width: width,
      start_cap: ast::CapStyle::Round,
      end_cap: ast::CapStyle::Round,
      join: ast::JoinStyle::Round,
      no_h_scale: false,
      no_v_scale: false,
      no_close: false,
      pixel_hinting: false,
      fill: ast::FillStyle::Solid(
        ast::fill_styles::Solid {
          color: ast::StraightSRgba8 {
            r: color.r,
            g: color.g,
            b: color.b,
            a: 255
          }
        }
      ),
    })
  )
}
